### PR TITLE
feat: MergeBranch function for GitLab wrapper

### PR DESCRIPTION
**Pull request by dragomirt**
This PR adds `mergeBranch` and `getBranchCommit` functions to GitLab's wrapper. **Still in testing**

### CODE REVIEW

File: lib/platform/gitlab/index.js [hunk 1]
```diff
@@ -335,8 +335,37 @@ async function deleteBranch(branchName, closePr = false) {
...
+async function mergeBranch(branchName, mergeType) {
+  logger.debug(`mergeBranch(${branchName}, ${mergeType})`);
+  // istanbul ignore if
+  if (config.pushProtection) {
+    logger.info(
+      { branchName, mergeType },
+      'Branch protection: Attempting to merge branch when push protection is enabled'
+    );
+  }
+  if (mergeType === 'branch-push') {
```
Comment (rarkins) [inline lib/platform/gitlab/index.js]: Now “branch” only

File: lib/platform/gitlab/index.js [hunk 2]
```diff
@@ -335,8 +335,37 @@ async function deleteBranch(branchName, closePr = false) {
   );
 }
 
-function mergeBranch() {
-  logger.warn('Unimplemented in GitLab: mergeBranch');
+async function mergeBranch(branchName, mergeType) {
+  logger.debug(`mergeBranch(${branchName}, ${mergeType})`);
+  // istanbul ignore if
+  if (config.pushProtection) {
```
Comment (rarkins) [inline lib/platform/gitlab/index.js]: We don't yet have this field in GitLab so this if statement can be removed

File: lib/platform/gitlab/index.js [hunk 3]
```diff
@@ -335,8 +335,37 @@ async function deleteBranch(branchName, closePr = false) {
...
+        `projects/${
+          config.repository
+        }/repository/commits/${sha}/cherry_pick?branch=${config.baseBranch}`
+      );
+    } catch (err) {
+      logger.warn(
+        expandError(err),
+        `Error pushing branch merge for ${branchName}`
+      );
+      throw new Error('branch-push failed');
```
Comment (rarkins) [inline lib/platform/gitlab/index.js]: We should align this with GitHub

File: lib/platform/gitlab/index.js [hunk 4]
```diff
@@ -335,8 +335,37 @@ async function deleteBranch(branchName, closePr = false) {
...
+  logger.debug(`mergeBranch(${branchName}, ${mergeType})`);
+  // istanbul ignore if
+  if (config.pushProtection) {
+    logger.info(
+      { branchName, mergeType },
+      'Branch protection: Attempting to merge branch when push protection is enabled'
+    );
+  }
+  if (mergeType === 'branch-push') {
+    const sha = await getBranchCommit(branchName);
```
Comment (rarkins) [inline lib/platform/gitlab/index.js]: Can't we just use the branch name instead of looking up `sha`?

File: lib/platform/gitlab/index.js [hunk 5]
```diff
@@ -353,6 +382,18 @@ async function getBranchLastCommitTime(branchName) {
   }
 }
 
+// Return the commit SHA for a branch
+async function getBranchCommit(branchName) {
```
Comment (rarkins) [inline lib/platform/gitlab/index.js]: We may still need this to determine if a branch is up to date with `master`

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/platform/gitlab/index.js [hunk 6]
```diff
@@ -23,7 +23,6 @@ module.exports = {
   deleteBranch,
   mergeBranch,
   getBranchLastCommitTime,
-  checkCompatibility,
```
Comment (dragomirt) [inline lib/platform/gitlab/index.js]: `checkCompatibility` should be exported because it's used in the tests.

**Comment by rarkins** [comment]
The exports of each platform need to be the same , so it cannot be exported and instead needs to be tested as part of mergeBranch

**Comment by rarkins** [comment]
@dragomirt I have removed the unnecessary code, could you update the tests?

**Review by rarkins — APPROVED** [review]

**Pull request merged**